src/AggregateFunctions/AggregateFunctionIf.cpp

```diff
@@ -136,28 +126,36 @@ class AggregateFunctionIfNullUnary final
         const IColumn * columns_param[] = {&column->getNestedColumn()};
 
         const IColumn * filter_column = columns[num_arguments - 1];
-        if (const ColumnNullable * nullable_column = typeid_cast<const ColumnNullable *>(filter_column))
-            filter_column = nullable_column->getNestedColumnPtr().get();
-        if constexpr (result_is_nullable)
+
+        const UInt8 * filter_values = nullptr;
+        const UInt8 * filter_null_map = nullptr;
+
+        if (filter_is_nullable)
         {
-            /// We need to check if there is work to do as otherwise setting the flag would be a mistake,
-            /// it would mean that the return value would be the default value of the nested type instead of NULL
-            if (is_all_zeros(assert_cast<const ColumnUInt8 *>(filter_column)->getData().data(), batch_size))
-                return;
+            const ColumnNullable * nullable_column = assert_cast<const ColumnNullable *>(filter_column);
+            filter_column = nullable_column->getNestedColumnPtr().get();
+            filter_null_map = nullable_column->getNullMapData().data();
         }
 
+        filter_values = assert_cast<const ColumnUInt8 *>(filter_column)->getData().data();
+
         /// Combine the 2 flag arrays so we can call a simplified version (one check vs 2)
         /// Note that now the null map will contain 0 if not null and not filtered, or 1 for null or filtered (or both)
-        const auto * filter_flags = assert_cast<const ColumnUInt8 *>(filter_column)->getData().data();
+
         auto final_nulls = std::make_unique<UInt8[]>(batch_size);
-        for (size_t i = 0; i < batch_size; ++i)
-            final_nulls[i] = (!!null_map[i]) | (!filter_flags[i]);
+
+        if (filter_null_map)
+            for (size_t i = 0; i < batch_size; ++i)
+                final_nulls[i] = (!!null_map[i]) | (!filter_values[i]) | (!!filter_null_map[i]);
+        else
+            for (size_t i = 0; i < batch_size; ++i)
+                final_nulls[i] = (!!null_map[i]) | (!filter_values[i]);
 
         this->nested_function->addBatchSinglePlaceNotNull(
             batch_size, this->nestedPlace(place), columns_param, final_nulls.get(), arena, -1);
 
         if constexpr (result_is_nullable)
-            if (!memoryIsByte(null_map, batch_size, 1))
+            if (!memoryIsByte(final_nulls.get(), batch_size, 1))
```

Comment:
Does it make sense to move the nested function call here? It seems like we could avoid it in some cases since we are going to ignore the result, right?

Reply:
58bb1ffecf

Reply:
❤️